src/features/tempRole/expireCheck.ts

```diff
@@ -0,0 +1,60 @@
+import { Guild, GuildMember, Role } from 'discord.js'
+
+import cron from 'node-cron'
+
+import { Environment } from '@/config'
+import { prisma } from '@/prisma'
+import { BotContext } from '@/types/BotContext'
+
+export async function expireCheck(botContext: BotContext) {
+  await cron.schedule(Environment.TIME_PERIOD_CRON, async () => {
+    // Check for expired temporary roles
+    const expiredTemporaryRoles = await prisma.tempRole.findMany({
+      where: {
+        expiresAt: {
+          lte: new Date(),
+        },
+      },
+      select: {
+        guildId: true,
+        userId: true,
+        roleId: true,
+      },
+    })
+    // Remove expired temporary roles from users
+    for (const expiredTemporaryRole of expiredTemporaryRoles) {
+      if (expiredTemporaryRole.guildId !== Environment.GUILD_ID) return
+      const { client } = botContext
+
+      // Get guild, member, and role
+      const guild = client.guilds.cache.get(
+        expiredTemporaryRole.guildId,
+      ) as Guild
+      const member = (await guild.members.fetch(
+        expiredTemporaryRole.userId,
+      )) as GuildMember
+      const role = guild.roles.cache.get(expiredTemporaryRole.roleId) as Role
+
+      try {
+        // Remove role from member
+        await member.roles.remove(role)
+
+        // Remove expired temporary roles from database
+        await prisma.tempRole.deleteMany({
+          where: {
+            expiresAt: {
+              lte: new Date(),
+            },
+          },
+        })
```

Comment:
`await member.roles.remove(role)` called take some time, that mean this 2 `new Date()` can have different value, there might be some tempRole that get skipped.

Also, `await prisma.tempRole.deleteMany` is inside loop, it will get call more than necessary. 

You should collect expiredTemporaryRole that successfully removed, then removed from database, something like this.

```ts
const removedTemporaryRole = []
// Remove expired temporary roles from users
for (const expiredTemporaryRole of expiredTemporaryRoles) {
    ...
    await member.roles.remove(role)
    removedTemporaryRole.push(expiredTemporaryRole.id)
    ...
}

await prisma.tempRole.deleteMany({
  where: {
    id: {
      in: removedTemporaryRole,
    },
  },
})
```